racket-src/body/type-of.rkt

```diff
@@ -91,19 +92,61 @@
   ;; Get the type of a field from a given variant of a given ADT,
   ;; substituting type parameters.
 
-  #:mode (field-ty I I I I I O)
-  #:contract (field-ty Γ AdtId Parameters VariantId FieldId Ty)
+  #:mode (field-tys I I I I O)
+  #:contract (field-tys Γ AdtId Parameters VariantId ((FieldId Ty) ...))
 
-  [(where (_ AdtId KindedVarIds where _ (_ ... (VariantId FieldDecls) _ ...)) (decl-of-adt Γ AdtId))
+  [(where (_ AdtId KindedVarIds where _ (_ ... (VariantId FieldDecls) _ ...)) (find-adt Γ AdtId))
    (where/error Substitution (create-substitution KindedVarIds Parameters))
-   (where (_ ... (FieldId Ty) _ ...) FieldDecls)
-   (where/error Ty_substituted (apply-substitution Substitution Ty))
+   (where ((FieldId Ty) ...) FieldDecls)
+   (where/error (Ty_substituted ...) ((apply-substitution Substitution Ty) ...))
    ----------------------------------
-   (field-ty Γ AdtId Parameters VariantId FieldId Ty_substituted)
+   (field-tys Γ AdtId Parameters VariantId ((FieldId Ty_substituted) ...))
    ]
 
   )
 
+(define-judgment-form
+  formality-body
+
+  #:mode (type-of/Constant I I O)
+  #:contract (type-of/Constant Γ Constant Ty)
+
+  [; integer
+   ; FIXME: choose proper int type
+   ------------------------------------------
+   (type-of/Constant Γ number (user-ty i32))]
+
+  [; true
+   ------------------------------------------
+   (type-of/Constant Γ true (user-ty bool))]
+
+  [; false
+   ------------------------------------------
+   (type-of/Constant Γ false (user-ty bool))]
+
+  [; function
+   (where/error (fn _ KindedVarIds (Ty_arg ...) -> Ty_ret _ _ _) (find-fn Γ FnId))
+   (where/error (KindedVarId_subst ..._n KindedVarId_other ...) KindedVarIds)
```

Comment:
Wait, what is this `_n` exactly? I don't understand why we want to take the first `N` generic parameters.

Reply:
I assume that the first `n` parameters of the function are early-bound and the rest are late-bound -- not sure if this is the right behavior. Thus a function pointer `(fn-ptr foo [T])` to `fn foo[(type T) (lifetime %a) (lifetime %b)] ((& %a (& %b ())) (& %b T)) -> (& %a T)` would get a type like `(∀[(lifetime %a) (lifetime %b)] (rigid-ty ...))`.